// react-spinners/dist/spinners/FadeLoader.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'prop-types', 'emotion', 'recompose'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('prop-types'), require('emotion'), require('recompose'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes, global.emotion, global.recompose);
    global.FadeLoader = mod.exports;
  }
})(this, function (exports, _react, _propTypes, _emotion, _recompose) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var fade = (0, _emotion.keyframes)([], [], function createEmotionStyledRules() {
    return [{
      '50%': {
        'opacity': '0.3'
      },
      '100%': {
        'opacity': '1'
      }
    }];
  });

  var Loader = function (_React$Component) {
    _inherits(Loader, _React$Component);

    function Loader() {
      var _ref;

      var _temp, _this, _ret;

      _classCallCheck(this, Loader);

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Loader.__proto__ || Object.getPrototypeOf(Loader)).call.apply(_ref, [this].concat(args))), _this), _this.style = function (i) {
        return (0, _emotion.css)([], [_this.props.height, _this.props.width, _this.props.margin, _this.props.color, _this.props.radius, fade, i * 0.12], function createEmotionStyledRules(x0, x1, x2, x3, x4, x5, x6) {
          return [{
            '': {
              'position': 'absolute',
              'height': x0 + 'px',
              'width': x1 + 'px',
              'margin': x2,
              'backgroundColor': x3,
              'borderRadius': x4,
              'WebkitTransition': '2s',
              'transition': '2s',
              'WebkitAnimationFillMode': '\'both\'',
              'animationFillMode': '\'both\'',
              'WebkitAnimation': x5 + ' 1.2s ' + x6 + 's infinite ease-in-out',
              'animation': x5 + ' 1.2s ' + x6 + 's infinite ease-in-out'
            }
          }];
        });
      }, _this.radius = 20, _this.quarter = _this.radius / 2 + _this.radius / 5.5, _this.wrapper = function () {
        return (0, _emotion.css)([], [_this.radius, _this.radius, _this.radius * 3, _this.radius * 3], function createEmotionStyledRules(x0, x1, x2, x3) {
          return [{
            '': {
              'position': 'relative',
              'fontSize': '0',
              'top': x0 + 'px',
              'left': x1 + 'px',
              'width': x2 + 'px',
              'height': x3 + 'px'
            }
          }];
        });
      }, _this.a = function () {
        return (0, _emotion.css)([_this.style(1)], [_this.radius], function createEmotionStyledRules(x0) {
          return [{
            'top': x0,
            'left': '0'
          }];
        });
      }, _this.b = function () {
        return (0, _emotion.css)([_this.style(2)], [_this.quarter, _this.quarter], function createEmotionStyledRules(x0, x1) {
          return [{
            'top': x0,
            'left': x1,
            'WebkitTransform': 'rotate(-45deg)',
            'transform': 'rotate(-45deg)'
          }];
        });
      }, _this.c = function () {
        return (0, _emotion.css)([_this.style(3)], [_this.radius], function createEmotionStyledRules(x0) {
          return [{
            'top': '0',
            'left': x0,
            'WebkitTransform': 'rotate(90deg)',
            'transform': 'rotate(90deg)'
          }];
        });
      }, _this.d = function () {
        return (0, _emotion.css)([_this.style(4)], [-_this.quarter, _this.quarter], function createEmotionStyledRules(x0, x1) {
          return [{
            'top': x0,
            'left': x1,
            'WebkitTransform': 'rotate(45deg)',
            'transform': 'rotate(45deg)'
          }];
        });
      }, _this.e = function () {
        return (0, _emotion.css)([_this.style(5)], [-_this.radius], function createEmotionStyledRules(x0) {
          return [{
            'top': x0,
            'left': '0'
          }];
        });
      }, _this.f = function () {
        return (0, _emotion.css)([_this.style(6)], [-_this.quarter, -_this.quarter], function createEmotionStyledRules(x0, x1) {
          return [{
            'top': x0,
            'left': x1,
            'WebkitTransform': 'rotate(-45deg)',
            'transform': 'rotate(-45deg)'
          }];
        });
      }, _this.g = function () {
        return (0, _emotion.css)([_this.style(7)], [-_this.radius], function createEmotionStyledRules(x0) {
          return [{
            'top': '0',
            'left': x0,
            'WebkitTransform': 'rotate(90deg)',
            'transform': 'rotate(90deg)'
          }];
        });
      }, _this.h = function () {
        return (0, _emotion.css)([_this.style(8)], [_this.quarter, -_this.quarter], function createEmotionStyledRules(x0, x1) {
          return [{
            'top': x0,
            'left': x1,
            'WebkitTransform': 'rotate(45deg)',
            'transform': 'rotate(45deg)'
          }];
        });
      }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Loader, [{
      key: 'render',
      value: function render() {
        return this.props.loading ? _react2.default.createElement(
          'div',
          { className: this.wrapper() },
          _react2.default.createElement('div', { className: this.a() }),
          _react2.default.createElement('div', { className: this.b() }),
          _react2.default.createElement('div', { className: this.c() }),
          _react2.default.createElement('div', { className: this.d() }),
          _react2.default.createElement('div', { className: this.e() }),
          _react2.default.createElement('div', { className: this.f() }),
          _react2.default.createElement('div', { className: this.g() }),
          _react2.default.createElement('div', { className: this.h() })
        ) : null;
      }
    }]);

    return Loader;
  }(_react2.default.Component);

  Loader.propTypes = {
    loading: _propTypes2.default.bool,
    color: _propTypes2.default.string,
    height: _propTypes2.default.number,
    width: _propTypes2.default.number,
    margin: _propTypes2.default.string,
    radius: _propTypes2.default.number
  };

  Loader.defaultProps = {
    loading: true,
    color: '#000000',
    height: 15,
    width: 5,
    margin: '2px',
    radius: 2
  };

  var Component = (0, _recompose.onlyUpdateForKeys)(['loading', 'color', 'height', 'width', 'margin', 'radius'])(Loader);
  Component.defaultProps = Loader.defaultProps;
  exports.default = Component;
});
